### ISSUE TITLE
Substituting expressions into sizes when possible

### ISSUE THREAD

**Issue by athas** [diku-dk/futhark#1945]
Consider this totally-not-contrived function:

```Futhark
def f (x: i64) =
  let y = x + 1
  in iota y
```

Currently this has type `(x: i64) -> ?[d₂].*[d₂]i64`, because when `y` goes out of scope, the size is made fully existential.  Why don't we substitute the definition of `y` instead?  Then `f` would have type `(x: i64) -> [x+1]i64`.

* This is possible whenever the definition is a valid size (meaning it does not consume anything), and it is of type `i64`.
* In other cases we can continue decaying to a full existential.
* This can result in duplication of code, e.g: `let y = x + 1 in replicate y (replicate y true)` (or even `iota (y*y*y)`).  Unclear whether this is mostly a usability problem (big size expressions are hard to read) or potentially an efficiency problem.

While the current behaviour is sound, it does discourage naming intermediate results, which is the opposite of what is normally considered good programming practice.

Note that I am not proposing substituting definitions of variables in scope during type checking.  E.g. this would still be ill-typed:

```Futhark
let y = x + 1
in zip (iota y) (iota (x+1))
```

Addressing *that* would require a deeper reworking of how the type checker works and raises lots of questions with unclear answers, while substituting variables that go out of scope would be a fairly local change.

**Comment by catvayor** [comment]
I don't think we want to do this all the time:
```futhark
def g (x: i64) : i64 = ... -- very expensive to calculate

def f (x: i64) =
  let y = g x
  in iota y
```
In such program, having an existential avoid us to calculate `g x` multiple times, as this calculation is completely hidden.
Doing this substitution would lead to calculate at least twice `g x` per call of `f`: in `f` and in the caller function.

Beside of this problem, which could be solved by adding an attribute ``do not substitute'' to the let binding, we still need to guarantee the result type depends only on arguments: what do we do for this program ?
```futhark
def f (x: i64) =
  let y = x + 1
  let z = y*y
  in iota z
```
We continue to substitute by the bounded expression ? when do we stop ? is it really meaningful to do this only for i64 at some point ?

> Note that I am not proposing substituting definitions of variables in scope during type checking. E.g. this would still be ill-typed:
> ```futhark
> let y = x + 1
> in zip (iota y) (iota (x+1))
> ```
> Addressing that would require a deeper reworking of how the type checker works and raises lots of questions with unclear answers, while substituting variables that go out of scope would be a fairly local change.

I'm not sure it would be *that* complicated to do, it certainly imply some working in the unification process. I think by giving the bindings to unification, it could unwrap bindings as needed when it comes to unify a variable to an expression. (It clearly needs some more thinking).

Also, by allowing this, we could propose something for our complex programs: when specifying the return type, we could do this unification with unwrapping of bindings to allow the programmer to decide which form the return type should take. Sadly, this doesn't solve the problem when not specifying the return type...

**Comment by athas** [comment]
> In such program, having an existential avoid us to calculate g x multiple times, as this calculation is completely hidden. Doing this substitution would lead to calculate at least twice g x per call of f: in f and in the caller function.

Good point, I hadn't thought of that.  Two counterpoints:

### The first

The programmer could still avoid this by assigning an explicit return type to `f`:

 ```Futhark
def f (x: i64) : ?[n].[n]i64 =
  iota (g x)
```

I think using a `let` binding as a cost-control mechanism is perilously close to sin.  It is very strange to me that `let tmp = g x in iota tmp` and `iota (g x)` should have different performance implications (which they will because of the different types assigned to a function with these expressions as its body).

### The second

I remain unconvinced that we really need to evaluate sizes occurring only in return types in advance.  If you put through program through monomorphisation:

```
def g (x: i64) : i64 = x+1 -- very expensive to calculate

def f (x: i64) =
  iota (g x)

def main x = f x
```

Then sure enough, the definition of `main` will contain an application of `g`.  However, the result is not actually used anywhere (not even in the instantiation of `f`).  Why do we have it?

### Other points

> We continue to substitute by the bounded expression ? when do we stop ? 

We never stop!  The type of that function would be `(x: i64) -> [(x+1)*(x+1)]i64`.  Yes, this can result in exponentially large expressions in sizes, so... just don't do that.  I don't think this is a big problem in practice in dependently typed languages.

**Comment by catvayor** [comment]
> Then sure enough, the definition of `main` will contain an application of `g`. However, the result is not actually used anywhere (not even in the instantiation of `f`). Why do we have it?

We have it because I had "unknown variable" internal errors for cases like
```futhark
def g (x: i64) : i64 = x+1 -- very expensive to calculate

def f (x: i64) =
  iota (g x)

def main x = 
  let t = f x
  in t
```
coming from the size in the type of `t`, `d<{g x}>`, which is written in the defunctionalised form of main:
```futhark
entry main (x: i64) : ?[d<{g x}>].*[d<{g x}>]i64 =
  let (d<{g x}>: i64) =
    g x
  let (t: [d<{g x}>]i64) =
    f x
  in t
```
(and it could also appear as an explicit size argument in some cases). So I just do the calculation every time, hoping for the optimizer to remove useless calculation.
I see that in the produced IR, as `f` acquired an existential return during monomorphisation, this size is returned by `f`, so the calculation is probably done only once in practice.
